### ISSUE TITLE
tracer flags in namelist

### ISSUE THREAD

**Issue by eclare108213** [CICE-Consortium/CICE#212]
Clean up redundant namelist flags, especially for bgc but the other tracers might have issues too.  This might be done in conjunction with reducing the memory footprint for tracers by allocating them based on the ones that are turned on.   This needs to be scoped/designed first...   and then also properly documented.

**Comment by eclare108213** [comment]
We don’t need both max_ntrcr (the previously compiled-in tracer array size) and ntrcr (the size actually used based on namelist inputs, which could be less than max_ntrcr).  Now that the code dynamically allocates variables, the sizes used to calculate max_ntrcr and the flags used to calculate ntrcr are ALL in namelist, and some of them are redundant.


         max_ntrcr =   1         & ! 1 = surface temperature
                   + nilyr       & ! ice salinity
                   + nilyr       & ! ice enthalpy
                   + nslyr       & ! snow enthalpy
                               !!!!! optional tracers:
                   + n_trage     & ! age
                   + n_trfy      & ! first-year area
                   + n_trlvl*2   & ! level/deformed ice
                   + n_trpnd*3   & ! ponds
                   + n_aero*4    & ! number of aerosols * 4 aero layers
                   + n_trbri     & ! brine height
                   + n_trbgcs*n_bgc           & ! skeletal layer BGC
                   + n_trzs  *n_trbri* nblyr  & ! zsalinity  (off if n_trbri=0)
                   + n_bgc*n_trbgcz*n_trbri*(nblyr+3) & ! zbgc (off if n_trbri=0)
                   + n_bgc*n_trbgcz           & ! mobile/stationary phase tracer
                   + 1             ! for unused tracer flags

Then count up ntrcr based on tr_[trcr] flags, e.g.

         if (tr_iage) then
             ntrcr = ntrcr + 1
             nt_iage = ntrcr   ! chronological ice age
         endif

ntrcr is used throughout the code, and I suspect it would be the natural choice to keep, instead of max_ntrcr.

For many tracers, the logical tr_[trcr] implies a fixed number of tracers are needed of that type.
For bgc, there’s a choice of how many to use.

Two options:

1)
For off/on choices, instead of setting n_[trcr] in namelist, use tr_[trcr] to set n_[trcr].
Where there is a choice of how many to use, use tr_[trcr] to turn off/on the parameterization and include n_[trcr] in the namelist for the desired number.

2)
For off/on choices, in namelist set n_[trcr]=0 for ‘off’, n_[trcr]=1 for ‘on’, and then in the code, set the logical flags appropriately or change the code to use n_[trcr] instead of the logical flags.
Where there is a choice of how many of that tracer type to use, set n_[trcr]=number desired.
This option seems more prone to user error, since it’s less clear which n_[trcr] are on/off and which allow larger sizes.

Either way, we could then get rid of code like this:

      if (tr_aero .and. n_aero==0) then
         if (my_task == master_task) then
            write(nu_diag,*) 'ERROR: aerosols activated but'
            write(nu_diag,*) 'ERROR:   not allocated in tracer array.'
            write(nu_diag,*) 'ERROR:   Activate in compilation script.'
         endif
         abort_flag = 9
      endif


tr_aero is problematic because it’s treated differently from the other tracers (4 layers, independent of the ice layers).  One way to fix this could be to separate it into 4 distinct tracers rather than having a uniquely indexed 4-tracer array.

**Comment by dabail10** [comment]
How do you want to split the duties on this one?

**Comment by eclare108213** [comment]
This has required splitting up the init_zbgc subroutine so that the ntrcr counting can be done before arrays in other parts of the code are allocated.  I'm almost done with that.  The remaining changes will be relatively small.  My time this week is really really overcommitted, so maybe I can submit a PR, some of you can test it, and I can provide instructions for what else to do.    It's easy for the non-BGC tracers.  Some of the BGC flags ought to be renamed and/or moved to other namelists, and we can decide whether that's worth doing.  Calling out @njeffery so she's aware of this conversation.

**Comment by apcraig** [comment]
I'm happy to help test or with any implementation at any point, just let me know.

**Comment by eclare108213** [comment]
Thanks.  It would be helpful if you can regression-test what I've done so far in a baseline suite.  I've been testing it in a slimmed-down suite that exercises the main bgc options, and that is BFB so I'm hopeful I haven't messed up other things. There are changes in icepack (branch tracer_init_icepack in my icepack fork) and in cice (branch tracer_init in my cice fork).  I haven't tried to pull the more recent changes (like the Icepack version update in CICE) into my forks -- I will try to do that next.  Then I can issue PRs -- I have to do one in Icepack and one in CICE, separately, right?  But the testing has to use them together.  This commit just has init_zbgc split up, without any changes to namelist yet, but this should set the stage for allocating tracers based on ntrcr instead of max_[trcr] and then cleaning up the namelist.

**Comment by apcraig** [comment]
OK, I'll start some testing.  Yes, we'll have to do two PRs.  But you don't have to link the two versions via the submodule until the last step.  Just continue to work in your sandbox on your CICE and Icepack branches, committing changes, updating from trunk, etc.  Then when you are ready, we'll want the Icepack PR first.  Once that's merged, you can update the Icepack submodule on your CICE branch and execute the CICE PR.

**Comment by apcraig** [comment]
I ran a full test suite on 4 compilers on conrad with the CICE tracer_init and Icepack tracer_init_branch versions.  All tests pass.  I compared against the CICE trunk version from Oct 28 and 90%+ of the cases did NOT pass the regression test.  It looks like this is expected based on the differences in the Icepack versions.  I may try to run again and compare against the current trunks.  At this point, I think things are fine.  We should retest again after the masters are merged onto the branches and before the PR is merged.
